main/src/ca/uwaterloo/flix/language/phase/EntryPoint.scala

```diff
@@ -18,264 +18,564 @@ package ca.uwaterloo.flix.language.phase
-    */
-  private def checkEntryPointResult(defn: TypedAst.Def, root: TypedAst.Root, traitEnv: TraitEnv)(implicit sctx: SharedContext, flix: Flix): Unit = defn match {
-    case TypedAst.Def(sym, TypedAst.Spec(_, _, _, _, _, declaredScheme, _, declaredEff, _, _), _, _) =>
-      val resultTpe = declaredScheme.base.arrowResultType
-      val unitSc = Scheme.generalize(Nil, Nil, Type.Unit, RigidityEnv.empty)
-      val resultSc = Scheme.generalize(Nil, Nil, resultTpe, RigidityEnv.empty)
-
-      // Check for IllegalEntryPointEffect
-      if (isBadEntryPointEffect(declaredEff)) {
-        val error = EntryPointError.IllegalEntryPointEff(sym, declaredEff, declaredEff.loc)
-        sctx.errors.add(error)
+    /** Returns `defn` without a test annotation. */
+    private def removeTestAnnotation(defn: TypedAst.Def): TypedAst.Def =
+      defn.copy(
+        spec = defn.spec.copy(
+          ann = defn.spec.ann.copy(
+            annotations = defn.spec.ann.annotations.filterNot(_.isInstanceOf[Annotation.Test])
+          )
+        )
+      )
+
+    /**
+      * Rules for exported functions - an exported function has:
+      *   - No type variables.
+      *   - An effect that is a subset of the primitive effects.
+      *   - Is not in the root namespace.
+      *   - Is `pub`.
+      *   - Has a name that is valid in Java.
+      *   - Has types that are valid in Java (not Flix types like `List[Int32]`).
+      */
+    private def visitExport(defn: TypedAst.Def)(implicit sctx: SharedContext, flix: Flix): TypedAst.Def = {
+      val errs = (checkNoTypeVariables(defn) match {
+        case Some(err) => List(err)
+        case None =>
+          // Only run these on functions without type variables.
+          checkPrimitiveEffect(defn).toList ++ checkJavaTypes(defn)
+      }) ++
+        checkNonRootNamespace(defn) ++
+        checkPub(defn) ++
+        checkValidJavaName(defn)
+      if (errs.isEmpty) {
+        defn
+      } else {
+        errs.foreach(sctx.errors.add)
+        removeExportAnnotation(defn)
+      }
+    }
+
+    /** Returns `defn` without a test annotation. */
+    private def removeExportAnnotation(defn: TypedAst.Def): TypedAst.Def =
+      defn.copy(
+        spec = defn.spec.copy(
+          ann = defn.spec.ann.copy(
+            annotations = defn.spec.ann.annotations.filterNot(_.isInstanceOf[Annotation.Export])
+          )
+        )
+      )
+
+    /**
+      * Returns an error if `defn` has type variables.
+      *
+      * If a function has no type variables in surface syntax we know that:
+      *   - Traits and trait constraints do not occur since their syntax is limited to `Trait[var]`.
+      *   - Associated types do not occur since their syntax is limited to `Trait.Assoc[var]`.
+      *   - Equality constraints do not occur since their syntax is limited to
+      *     `Trait.Assoc[var] ~ type`
+      */
+    private def checkNoTypeVariables(defn: TypedAst.Def): Option[EntryPointError] = {
```

Comment:
To what extent could all these `check*` things be turned into functions with a side-effect of modifying the context?

Reply:
they could all be I suppose, but i like to keep them sideffect free when possible

The functions that don't use shared context is because doing either is easy